[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119107
* __->__ #118438

Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

williamwen42 committed:
[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/cache_entry.cpp:
```diff
@@ -0,0 +1,111 @@
+  Py_TYPE(e)->tp_free((PyObject*)e);
+}
+
+PyTypeObject CacheEntryType = {
```
Consider rewriting this to use pybind.

jansel commented on torch/csrc/dynamo/cache_entry.cpp:
```diff
@@ -0,0 +1,111 @@
+CacheEntry* create_cache_entry(
+    CacheEntry* next,
+    PyObject* guarded_code) {
+  PyObject* check_fn = PyObject_GetAttrString(guarded_code, "check_fn"); // new reference
+  PyCodeObject* code = (PyCodeObject*)PyObject_GetAttrString(guarded_code, "code"); // new reference
+
+  // equivalent to CacheEntry(check_fn, code, next) in Python
+  PyObject* args = Py_BuildValue("OOO", check_fn, code, next);
+  CacheEntry* e = (CacheEntry*)PyObject_CallObject((PyObject*)&CacheEntryType, args); // new reference
```
Missing error checks on all these calls.

jansel commented on torch/csrc/dynamo/cache_entry.cpp:
```diff
@@ -0,0 +1,111 @@
+
+  if (!ret) return -1;
+
+  if (check_fn) {
```
When will this be false?  Shouldn't that be an error?

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/cache_entry.h:
```diff
@@ -0,0 +1,66 @@
+
+typedef struct CacheEntry {
+  // check the guards: lambda: <locals of user function>: bool
+  py::object check_fn;
```
Have you benchmarked this?  This (and code) are the only part on the critical path, so want to much sure making it a py::object doesn't slow things down.

You can use ./benchmarks/dynamo/microbenchmarks/overheads.py

williamwen42 replied:
My results are:
```
pre changes
eager    6.4us (warmup=0.0s)
compiled 15.0us (warmup=8.3s)

post changes
eager    6.2us (warmup=0.0s)
compiled 14.9us (warmup=9.1s)
```

jansel replied:
Yes seems fine, so just need to address test failures.

jansel reviewed (changes requested):
Test failures?

Other than minor comment and test failures looks good.

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on .lintrunner.toml:
```diff
@@ -463,6 +463,7 @@ include_patterns = [
 ]
 exclude_patterns = [
     'torch/csrc/utils/pybind.h',
+    'torch/csrc/dynamo/**/*',
```
Why?  Can we just fix the lints?

williamwen42 replied:
I was experimenting to see if including `torch/csrc/utils/pybind.h` over `pybind/pybind11.h` was causing issues in CI. I don't think this is the case - I've reverted the lintrunner settings.

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor CacheEntry and ExtraState to eval_frame.c to C++"


Part of implementing CacheEntry invalidation to fix https://github.com/pytorch/pytorch/issues/112090.

Changes:
- Move CacheEntry and ExtraState to C++
- Use pybind to control reference counting
- Use std::list instead of manually implementing a linked list

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge
